오디오 시스템 구현, 어깨 적용

daminoworld opened the pull request:
# 이슈
- close #109
# 작업 사항
- 이전 PR에서 subscribeSceneEvent 메서드를 지웠어야 했는데 실수로 SetHandRocketEntity 메서드를 삭제햐여 수정(1.0 릴리즈에 반영 됨)

### SoundEffectHelper 구현
- 효과음 오디오 파일을 RCP에서 삭제하고, AudioResources 폴더에 추가하여 사용하는것으로 수정하였습니다.
- SoundEffectProtocol을 만들고 extension을 통해 RawRepresentable을 채택한 타입에 대해서만 적용(RawRepresentable은 열거형이나 구조체가 원시 값을 가질 수 있게 하는 프로토콜)
- RawRepresentable을 채택한 타입에서 원시 값이 String 타입인 경우에만 extension이 적용
- SoundHelper 클래스를 SoundEffectProtocol을 준수하는 제네릭 타입을 받아서 사용
- init시 제네릭 타입으로 받은 타입 enum을 통해 AudioFileResources를 로드하고, 로드 한값을 loadedSounds 딕셔너리에 추가
- 재생시, Spatial audio를 위해 SpatialAudioComponent를 추가했고, effect 타입과 엔터티를 받아서 로드한 resource를 가져와 재생

### 사용방법
1. SoundEffectHelper 파일 아래에 부위별 enum을 만들었고, 해당 enum의 케이스 이름을 AudioResources에 넣을 효과음 이름과 같게 해줍니다.
2. 부위 파일 viewModel에 var soundHelper = SoundEffectHelper<`부위별enum 타입`>() 으로 soundHelper를 정의하고
3. soundHelper.playSound(.`부위별enum의케이스선택`, on: `적용엔터티`)로 오디오를 실행 시켜주시면 됩니다.

# 고민 or 참고 사항 (선택)
- 모든 부위에서 범용적으로 쉽게 쓸 수 있게 만들기는 했지만 enum케이스와 오디오 파일 명을 일치시켜서 작성해줘야 하는 번거로움이 있기는 합니다. 더 좋은 의견있으시면 반영하겠습니다~!
- 어깨는 RCP에서 오디오 파일과 SpatialAudio엔터티 다 삭제해서 시원하네요

### 코드 리뷰 반영
- 기존에 SoundEffectProtocol, SoundEffectHelper에서 playSound라는 같은 이름으로 다른 동작을 하는 메서드를 사용하여 가독성의 문제가 될 여지가 있어 Protocol에서는 playSound를 빼고, AudioEffect라는 이름으로 변경하고, 모든 playSound의 동작을 SoundEffectHelper의 playSound에서 하도록 변경

- loadSounds() 가 실행되고 playSound가 실행될 수 있도록 DispatchSemaphore 를 사용하여 실행 흐름 제어
- withThrowingTaskGroup를 사용하여 오디오 파일을 병렬적으로 불러오도록 구현
- isSpatial을 playSound 파라미터에 추가하여 SpatialAudioComponent 추가를 선택할 할 수 있도록 수정(default 는 true)

# 스크린샷 (선택)

daminoworld committed:
Chore: Replace Shoulder timer audio

daminoworld committed:
Fix: Recover SetHandRocketEntity

- 이전 PR에서 subscribeSceneEvent 메서드를 지웠어야 했는데 실수로 SetHandRocketEntity 메서드를 삭제햐여 수정

jsangyun committed:
Move: AudioResources 폴더 구조화

jsangyun committed:
Add: 눈 스트레칭 오디오 mp3 파일들 추가

daminoworld committed:
Del: Remove non used scenes

daminoworld committed:
Chore: Add ShoulderEntrySound to RocketScene

daminoworld committed:
Chore: Add play entry rocket audio logic

- 오디오 파일 구조 수정 전 임시로 오디오 재생 로직 추가

daminoworld committed:
Add: Add Shoulder Stretching audio files

daminoworld committed:
Feat: Implement SoundEffectHelper

- SoundEffectProtocol을 만들고
- 확장은 RawRepresentable을 채택한 타입에 대해서만 적용(RawRepresentable은 열거형이나 구조체가 원시 값을 가질 수 있게 하는 프로토콜)
- RawRepresentable을 채택한 타입에서 원시 값이 String 타입인 경우에만 확장이 적용
- SoundHelper 클래스를 SoundEffectProtocol을 준수하는 제네릭 타입을 받아서 사용
- init시 제네릭 타입으로 받은 타입 enum을 통해 AudioFileResources를 로드하고, 로드 한값을 loadedSounds 딕셔너리에 추가
- 재생시, effect 타입과 엔터티를 받아서 로드한 resource를 가져와 재생

daminoworld committed:
Feat: Initialize soundHelper in shoulder tutorial viewModel

daminoworld committed:
Move: Move SoundEffectHelper directory

daminoworld committed:
Feat: Make stars to get star case count

daminoworld committed:
Refactor: Replace audio logic with soundEffectHelper

daminoworld committed:
Del: Remove audio file and entity from RCP

MartyTheGout commented on APPRO/APPRO/Views/Tutorial/Shoulder/ShoulderStretchingTutorialView.swift:
```diff
@@ -111,22 +109,29 @@ struct ShoulderStretchingTutorialView: View {
-        Task {
-            await viewModel.playSpatialAudio(collidedModelEntity)
+        
+        debugPrint("디버그1", collidedModelEntity.name)
```
debugPrint는 삭제하면 좋을것같습니다!

daminoworld replied:
삭제하였습니다!

MartyTheGout commented on APPRO/APPRO/Modules/SoundEffectHelper.swift:
```diff
@@ -0,0 +1,82 @@
+}
+
+extension SoundEffectProtocol where Self: RawRepresentable, Self.RawValue == String {
+    var fileName: String { rawValue }
```
무매한 질문일수도 있겠습니다만, 

13 열의 var fileName: String { get } 과 18열의 var fileName: String { rawValue } 는 다르게 작동하는 것인가요? 

순수무매질문입니다호호

daminoworld replied:
Protocol의 fileName은 SoundEffectProtocol를 채택한 타입이 fileName을 구현하도록 강제하는데
extension에 fileName을 rawValue로 제공함으로써  Self.RawValue == String , 즉 rawValue가 String 타입일때는
즉, SoundEffectProtocol을 채택한 타입이 String을 채택한 경우
 fileName을 구현할 필요없이 RawValue로 fileName이 작성됩니다.

MartyTheGout commented on APPRO/APPRO/Modules/SoundEffectHelper.swift:
```diff
@@ -0,0 +1,82 @@
+extension SoundEffectProtocol where Self: RawRepresentable, Self.RawValue == String {
+    var fileName: String { rawValue }
+    
+    func playSound(on entity: Entity, resource: AudioFileResource) {
```
오 재미있습니다! 이전에 위니의 코드에서 보았던 모습과 유사하게 보이는데요. 
(또한 순수무매질문입니다) 

그 유사한 모습이 어떤 모습이냐하면, 
확장관계에 있거나(위니의 경우) , 서로다른 클래스에 있는(비전의 경우) 함수들을 같은 이름으로 구성하는 것입니다. 

<비전>
SoundEffectHelper_playSound() 
 ㄴ SoundEffectProtocol_playSound() 
 
 <위니> 
 EyeStretchingRingEntity_playAudio()
  ㄴ Exteded Entity_ async playAudio()
   ㄴ Entity's original _ playAudio() 
   
   처음 읽었기 때문에, 더욱 낯설게 느껴진 점도 있습니다만, 각각의 함수들은 구체적으로 들어갔을 때에는 서로 무엇을 하는지, 또한 지게되는 책임범위에 대해서도 조금씩 다른것으로 보이는데, 이들을 동일한 함수명을 지었을 때에 모두 같은 무언가를 하며 같은 책임범위를 지는 함수인것으로 보이는듯? 하게 느껴졌습니다. 
   
 물론 두분이 완벽히 동일한 쓰임새로 작성하시는 것은 아니지만, 다년간의 경험을 다진 개발자들에게 녹아있는 컨벤션과 같은 것들이 아닐까, 위의 이해와는 다르게 다른 이해가 필요하기 때문에 같은 이름으로 작성된 것은 또 아닐까 라는 생각을 하며 글을 남깁니다. 
 
 저의 추측이 맞을까요? 호호호호

daminoworld replied:
좋은 지적 감사합니다. 아래와 같이 코드 수정하였으니 확인부탁드립니다~
- 기존에 SoundEffectProtocol, SoundEffectHelper에서 playSound라는 같은 이름으로 다른 동작을 하는 메서드를 사용하여 가독성의 문제가 될 여지가 있어 Protocol에서는 playSound를 빼고, AudioEffect라는 이름으로 변경하고, 모든 playSound의 동작을 SoundEffectHelper의 playSound에서 하도록 변경

MartyTheGout reviewed (approved):
LGTM

RCP의 파일정리도 정리지만, 각각의 뷰모델 역할을 하는 계층에서 비슷한 역할을 하는 함수를 중복적으로 작성하고 또 사용하는것을 정리할 수 있는, 앱전반에 사용되는 스페이셜오디오 기능들을 통합할 수 있는 멋진 코드 같습니다! 

늦은 시간 까지 고생많으셨습니다!

daminoworld committed:
Chore: Remove debugPrint

daminoworld committed:
Refactor: Remove playSound from protocol

- 기존에 SoundEffectProtocol, SoundEffectHelper에서 playSound라는 같은 이름으로 다른 동작을 하는 메서드를 사용하여 가독성의 문제가 될 여지가 있어 Protocol에서는 playSound를 빼고, AudioEffect라는 이름으로 변경하고, 모든 playSound의 동작을 SoundEffectHelper의 playSound에서 하도록 변경

jsangyun commented on APPRO/APPRO/Modules/SoundEffectHelper.swift:
```diff
@@ -0,0 +1,76 @@
+    init() {
+        Task {
+            await loadSounds()
+        }
```
생성자 내부에서 Task를 통해 비동기로 `loadSounds`를 실행하면 소리들이 전부 불러와지지 않은 상태에서`SoundEffectHelper`가 호출될 가능성은 없는걸까요?

daminoworld replied:
가능성이 있어서 추가적으로 수정하여 커밋하였습니다! 감사합니다

jsangyun commented on APPRO/APPRO/Modules/SoundEffectHelper.swift:
```diff
@@ -0,0 +1,76 @@
+            } catch {
+                debugPrint("Failed to load sound: \(effect.fileName), \(error)")
+            }
+        }
```
`withThrowingTaskGroup`을 통해 사운드를 불러오는 작업들을 병렬적으로 수행할 수 있도록 하는 것도 좋을 것 같습니다!

daminoworld replied:
의견 반영하여 코드 수정하였습니다 👍

jsangyun commented on APPRO/APPRO/Modules/SoundEffectHelper.swift:
```diff
@@ -0,0 +1,76 @@
+            let audioController = entity.prepareAudio(resource)
+            audioController.play()
+        }
+    }
```
경우에 따라서 `AudioPlaybackController`를 외부에서 관리해야할 수도 있을 것 같아서 `audioController`를 리턴해주되 `@discardableResult`를 통해 반환값을 무시할 수 있도록 해보는 것도 좋을 것 같습니다!

daminoworld replied:
요 부분은 필요할 때 수정하면 될것 같아서 일단 유지 하였습니다.

jsangyun commented on APPRO/APPRO/Modules/SoundEffectHelper.swift:
```diff
@@ -0,0 +1,76 @@
+                debugPrint("Sound not loaded: \(effect.fileName)")
+                return
+            }
+            entity.components.set(SpatialAudioComponent())
```
`SoundEffectHelper`가 Entity의 Component를 설정하는 책임도 가져야하는가? 에 대해서 고민해보면 좋을 것 같네요!
어쩌면 공간 음향을 쓰지 않는 Entity도 있지 않을까요?

daminoworld replied:
쓰지 않는 엔터티가 있을 수 있어 우선 파라미터로 선택할 수 있도록 수정하였습니다!

jsangyun reviewed (approved):
고생하셨습니다!
범용적으로 쓰일 수 있도록 고민하신게 보이는 코드네요~ 배워갑니다 👍👍
눈 스트레칭에도 적용해보면서 추가적인 의견이나 궁금증이 생기면 말씀드릴게용~

daminoworld committed:
Refactor: Apply code reviews

- loadSounds() 가 실행되고 playSound가 실행될 수 있도록 DispatchSemaphore 를 사용하여 실행 흐름 제어
- withThrowingTaskGroup를 사용하여 오디오 파일을 병렬적으로 불러오도록 구현
- isSpatial을 playSound 파라미터에 추가하여 SpatialAudioComponent 추가를 선택할 할 수 있도록 수정(default 는 true)
